pints/_mcmc/_monomial_gamma_hamiltonian.py

```diff
@@ -0,0 +1,515 @@
+#
+# Monomial gamma HMC MCMC method
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import pints
+import numpy as np
+from scipy import integrate
+from scipy import interpolate
+
+
+class MonomialGammaHamiltonianMCMC(pints.SingleChainMCMC):
+    r"""
+    Implements Monomial Gamma HMC as described in [1] - a generalisation
+    of HMC as described in [2] - involving a non-physical kinetic energy term.
+
+    Uses a physical analogy of a particle moving across a landscape under
+    Hamiltonian dynamics to aid efficient exploration of parameter space.
+    Introduces an auxilary variable -- the momentum (``p_i``) of a particle
+    moving in dimension ``i`` of negative log posterior space -- which
+    supplements the position (``q_i``) of the particle in parameter space. The
+    particle's motion is dictated by solutions to Hamilton's equations,
+
+    .. math::
+        dq_i/dt =   \partial H/\partial p_i,
+        dp_i/dt = - \partial H/\partial q_i.
+
+    The Hamiltonian is given by,
+
+    .. math::
+        H(q,p) =       U(q)       +        K(p)
+               = -log(p(q|X)p(q)) +
+                 \Sigma_{i=1}^{d} (
+                 -g(p_i) + (2/c) \text{log}(1 + \text{exp}(cg(p_i))))
+
+    where ``d`` is the dimensionality of model, ``U`` is the potential
+    energy and ``K`` is the kinetic energy term. Note the kinetic energy
+    is the 'soft' version described in [1], where,
+
+    .. math::
+        g(p_i) = (1 / m_i) \text{sign}|p_i|^{1 / a}
+
+    To numerically integrate Hamilton's equations, it is essential to use a
+    sympletic discretisation routine, of which the most typical approach is
+    the leapfrog method,
+
+    .. math::
+        p_i(t + \epsilon/2) &= p_i(t) - (\epsilon/2) dU(q_i)/ dq_i\\
+        q_i(t + \epsilon) &= q_i(t) + \epsilon d K(p_i(t + \epsilon/2))/dp_i\\
+
+    The derivative of the soft kinetic energy term is given by,
+
+    .. math::
+        d K(p_i)/dp_i =
+        |p_i|^{-1 + 1 / a}\text{sign}(p_i) \times
+        \text{tanh}(c|p_i|^{1/a}\text{sign}(p_i) / {2 m_i}) / {a m_i}
+
+    In particular, the algorithm we implement follows eqs. (4.14)-(4.16) in
+    [2], since we allow different epsilon according to dimension.
+
+    [1] Towards Unifying Hamiltonian Monte Carlo and Slice Sampling
+    Yizhe Zhang, Xiangyu Wang, Changyou Chen, Ricardo Henao, Kai Fan,
+    Lawrence Cari.
+    [2] MCMC using Hamiltonian dynamics
```

Comment:
This doesn't render nicely, so maybe add a blank line before the [2]?
We'll have to decide on some system for this some time (e.g. when we're changing the style of the docs #849 )